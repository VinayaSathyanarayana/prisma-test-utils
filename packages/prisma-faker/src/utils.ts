/**
 * Fallbacks to default value if expression returns undefined or null.
 *
 * @param fallback
 */
export function withDefault<T>(fallback: T, val: T | undefined): T {
  if (val === undefined || val === null) return fallback
  return val
}
