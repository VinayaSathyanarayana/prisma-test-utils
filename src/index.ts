export { seed } from './seed'
export * from './types'
