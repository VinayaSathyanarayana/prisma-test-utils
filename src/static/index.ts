export {
  getMySQLPool,
  getPostgreSQLPool,
  getSQLitePool,
  MySQLPoolOptions,
  PostgreSQLPoolOptions,
  SQLitePoolOptions,
} from './pool'
export { getSeed } from './seed'
export * from './types'
