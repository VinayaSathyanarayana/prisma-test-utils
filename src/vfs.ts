import * as fs from 'fs'
import * as path from 'path'
import {
  CompilerOptions,
  createCompilerHost,
  createProgram,
  createSourceFile,
  ModuleKind,
  ScriptTarget,
} from 'typescript'
import { promisify } from 'util'

const mkdir = promisify(fs.mkdir)
const writeFile = promisify(fs.writeFile)

/**
 * Portrays the file system as a tree.
 */
export type VirtualFS = {
  [path: string]: string
}

/**
 * Searches virtual file system for TS files and compiles them.
 * Leaves non-TS files untacked.
 *
 * @param vfs
 */
export async function compileVFS(
  vfs: VirtualFS,
  options: CompilerOptions,
): Promise<VirtualFS> {
  /* Files */

  const tsFiles = Object.keys(vfs).filter(file => file.endsWith('.ts'))
  const nonTsFiles = Object.keys(vfs).filter(file => !file.endsWith('.ts'))

  const compiledVFS: VirtualFS = nonTsFiles.reduce(
    (acc, file) => ({
      ...acc,
      [file]: vfs[file],
    }),
    {},
  )

  /* Compiler Configuration */

  const compilerHost = createCompilerHost(options)
  const { getSourceFile, fileExists, readFile } = compilerHost

  compilerHost.getSourceFile = (fileName, target) => {
    /**
     * Load from the VFS or the system.
     */
    if (Object.hasOwnProperty(fileName)) {
      return createSourceFile(fileName, vfs[fileName], target, true)
    } else {
      return getSourceFile(fileName, target)
    }
  }

  compilerHost.fileExists = fileName => {
    return fileExists(fileName)
  }

  compilerHost.readFile = fileName => {
    if (vfs.hasOwnProperty(fileName)) {
      return vfs[fileName]
    } else {
      return readFile(fileName)
    }
  }

  compilerHost.writeFile = (fileName, data) => {
    compiledVFS[fileName] = data
  }

  /* Run the compiler */

  const program = createProgram(tsFiles, options, compilerHost)
  program.emit()

  return compiledVFS
}

/**
 * Writes virtual file system representation to the file system.
 *
 * @param vfs
 */
export async function writeToFS(root: string, vfs: VirtualFS): Promise<void> {
  const actions = Object.keys(vfs).map(async filePath => {
    await mkdir(path.dirname(path.join(root, filePath)), { recursive: true })
    await writeFile(path.join(root, filePath), vfs[filePath])
  })

  try {
    await Promise.all(actions)
  } catch (err) {
    throw err
  }
}
