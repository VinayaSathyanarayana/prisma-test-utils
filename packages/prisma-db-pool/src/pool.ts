import { LiftEngine, DataSource } from '@prisma/lift'
import { DMMF } from '@prisma/photon/runtime/dmmf-types'
import * as fs from 'fs'
import * as os from 'os'
import * as path from 'path'
import pify from 'pify'
import { withDefault } from './utils'

const writeFile = pify(fs.writeFile)
const makeDir = pify(fs.mkdir)

export type PoolDefinition = {
  dmmf: DMMF.Document
  pool: {
    min?: number
    max: number
  }
  cwd?: string
}

export type DBInstance = {
  cwd: string
  datamodel: string
}

export class Pool {
  private dmmf: DMMF.Document
  private cwd: string
  private dbs: {
    booting: string[]
    idle: DBInstance[]
    busy: DBInstance[]
  }
  private waiters: Waiter[]
  private capacity: number

  constructor(definition: PoolDefinition) {
    this.dmmf = definition.dmmf
    this.cwd = withDefault(process.cwd(), definition.cwd)
    this.dbs = {
      booting: [],
      idle: [],
      busy: [],
    }
    this.waiters = []
    this.capacity = definition.pool.max

    /* Start init */

    for (let index = 0; index < withDefault(0, definition.pool.min); index++) {
      this.createDBInstance().then(db => {
        this.dbs.idle = [...this.dbs.idle, db]
      })
    }
  }

  /**
   * Creates a new DB instances and lifts a migration.
   */
  private async createDBInstance(): Promise<DBInstance> {
    try {
      /* Constants */

      const id = Math.random()
        .toString(36)
        .slice(2)
      const tmpDir = os.tmpdir()
      const datasources: DataSource[] = [
        {
          name: 'db',
          connectorType: 'sqlite',
          url: `file:${tmpDir}`,
          config: {},
        },
      ]
      const dbFolder = path.join(tmpDir, './db/')

      /* Occupy resource pool. */
      this.dbs.booting = [...this.dbs.booting, id]

      /* Migrate Datamodel */
      const lift = new LiftEngine({
        projectDir: this.cwd,
      })

      const { datamodel } = await lift.convertDmmfToDml({
        dmmf: JSON.stringify(this.dmmf.datamodel),
        dataSources: datasources,
      })

      /* Release resource pool. */
      this.dbs.booting = this.dbs.booting.filter(dbId => dbId !== id)

      return {
        cwd: tmpDir,
        datamodel: datamodel,
      }
    } catch (err) {
      throw err
    }
  }

  /**
   * Returns a Photon instance from the pool once made available, and cleans
   * and populates the instance if necessary.
   *
   * The returned Photon instance includes the db identifier used for
   * the database releasing.
   *
   * @param opts
   */
  async getDBInstance(): Promise<DBInstance> {
    /**
     * Check whether any of the instances is available and creates a new one
     * if the pool limit is not yet reached, otherwise waits for the available
     * instance.
     */
    if (this.dbs.idle.length > 0) {
      /* If available take a resource and return it right away. */
      const [db, ...remaining] = this.dbs.idle
      this.dbs.idle = remaining
      this.dbs.busy = [...this.dbs.busy, db]

      return db
    } else if (
      this.dbs.idle.length + this.dbs.busy.length + this.dbs.booting.length <
      this.capacity
    ) {
      /* If full capacity is not yet reached create new instance. */
      const db = await this.createDBInstance()
      this.dbs.busy = [...this.dbs.busy, db]

      return db
    } else {
      /* Add to the line. */
      const waiter = new Waiter()
      this.waiters = [...this.waiters, waiter]

      return waiter.wait()
    }
  }

  /**
   * Makes the instance available.
   */
  async releaseDBInstance(db: DBInstance): Promise<void> {
    /**
     * Find the busy instance, remove that instance and give
     * it to the next in line.
     */
    const instance = this.dbs.busy.find(bdb => bdb.cwd === db.cwd)
    this.dbs.busy.filter(bdb => bdb.cwd !== db.cwd)

    /* Add to the next in line or make idle. */
    if (this.waiters.length > 0) {
      const [waiter, ...remainingWaiters] = this.waiters
      this.waiters = remainingWaiters
      waiter.allocate(instance)
    } else {
      this.dbs.idle = [...this.dbs.idle, instance]
    }
  }

  /**
   * Drains the pool by deleting all instances.
   */
  async drain(): Promise<void> {
    // TODO
  }
}

class Waiter {
  private promise: Promise<DBInstance>
  private resolve: (dbi: DBInstance) => void

  constructor() {
    this.promise = new Promise((resolve, reject) => {
      this.resolve = resolve
    })
  }

  wait(): Promise<DBInstance> {
    return this.promise
  }

  allocate(instance: DBInstance): void {
    this.resolve(instance)
  }
}
