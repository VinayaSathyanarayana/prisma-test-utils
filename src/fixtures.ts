import { DMMF } from '@prisma/dmmf'
import Model from '@prisma/dmmf/dist/Model'
import * as faker from 'faker'
import * as _ from 'lodash'
import { Dictionary } from 'lodash'
// import { readPrismaYml, findDatamodelAndComputeSchema } from './datamodel'
import { Faker, FakerBag, FakerSchema, ID } from './types'
import { withDefault } from './utils'

/**
 * Seed the database with mock data.
 *
 * @param fakerSchemaDefinition
 * @param opts
 */
export function seed(
  dmmf: DMMF,
  fakerSchemaDefinition?: Faker,
  opts: { seed?: number; silent?: boolean } = { seed: 42, silent: false },
): object[] | Promise<object[]> {
  /* FakerBag, Defaults */
  const bag: FakerBag = { faker }
  const DEFAULT_AMOUNT = 5
  // const DEFAULT_CONSTRAINT = bag.constraints.atMax(5)

  /* Prisma and Model evaluation */
  // const prisma = readPrismaYml()
  // const dmmf = findDatamodelAndComputeSchema(prisma.configPath, prisma.config)

  const fakerSchema = fakerSchemaDefinition(bag)

  /* Fixture calculations */
  const orders: Order[] = getOrdersFromDMMF(dmmf)
  const steps: Step[] = getStepsFromOrders(orders)
  const fixtures: Fixture[] = getFixturesFromSteps(fakerSchema, steps)

  // TODO: seeding

  if (opts.silent) {
    return fixtures
  } else {
    return Promise.resolve(fixtures)
  }

  /**
   * The Core logic
   * 1. Create orders from models. Orders tell how many instances of a model
   *  should exist in the end.
   * 2. Create steps from orders. Steps cronologically order the creation of the actual
   *  instances to enable relations.
   * 3. Convert steps to virual data instances - fixtures, to calculate relation ids.
   * 4. Apply mock generation functions to obtain actual data, relations are represented as lists
   *  of strings.
   */

  /**
   * Prisma Faker uses intermediate step between model-faker definition
   * and database seeding. Fixture is a virtual data unit used to describe
   * future data and calculate relations.
   */
  interface Order {
    model: Model
    amount: number
    relations: Dictionary<{ min: number; max: number }>
  }

  /**
   * Note the `relationTo` field; it defines the direction of a relation.
   * This helps with the execution process calculation. If the relation is
   * pointing towards the model, we shouldn't create it since the cyclic complement
   * will implement it.
   */
  interface Step {
    order: number // the creation order of a step, starts with 0
    model: Model
    amount: number // number of instances created in this step
    runningNumber: number // specifies the total number of all instances
    relations: Dictionary<{
      type: '1-1' | '1-N' | 'M-N'
      relationTo: string // determines the direction of relation
      amount: number
    }>
  }

  type FixtureData = Dictionary<
    ID | string | number | boolean | ID[] | string[] | number[] | boolean[]
  >

  /**
   * Represents the virtual unit.
   */
  interface Fixture {
    order: number // starts with 0
    id: string
    model: Model
    data: FixtureData
  }

  /* Helper functions */

  /**
   * Converts dmmf to fixtures. Fixtures represent the summary of a mock requirements
   * for particular model.
   *
   * @param dmmf
   */
  function getOrdersFromDMMF(dmmf: DMMF): Order[] {
    return dmmf.datamodel.models.map(model => {
      const fakerModel = withDefault(
        {
          amount: DEFAULT_AMOUNT,
          factory: undefined,
        },
        fakerSchema[model.name],
      )

      /* Generate relations based on provided restrictions. */
      const relations: Order['relations'] = model.fields
        .filter(f => f.isRelation())
        .reduce((acc, field) => {
          const fakerField = fakerModel.factory[field.name]

          switch (typeof fakerField) {
            case 'object': {
              const min = withDefault(0, fakerField.min)
              const max = withDefault(min, fakerField.max)

              return { [field.name]: { min, max } }
            }
            default: {
              throw new Error(`Expected a relation got ${typeof fakerField}`)
            }
          }
        }, {})

      return {
        model: model,
        amount: fakerModel.amount,
        relations: relations,
      }
    })
  }

  /**
   * Coverts fixtures to steps. Steps represent an ordered entity that specifies the creation
   * of a virtual data.
   *
   * Creates a pool of available instances and validates relation constraints.
   *
   * @param fixtures
   */
  function getStepsFromOrders(orders: Order[]): Step[] {
    type Pool = Dictionary<{
      model: Model
      remainingUnits: number
      allUnits: number
    }>

    const pool: Pool = orders.reduce(
      (acc, order) => ({
        ...acc,
        [order.model.name]: {
          model: order.model,
          remainingUnits: order.amount,
        },
      }),
      {},
    )

    /**
     * The sort function functionally implements topological sort algorithm
     * by making sure all relations have been defined prior to the inclusion of
     * an order in the chain.
     */
    function sort(
      remainingOrders: Order[],
      sortedSteps: Step[],
      pool: Pool,
    ): Step[] {
      switch (remainingOrders.length) {
        case 0: {
          return []
        }
        case 1: {
          const [o] = remainingOrders

          /* Checks if the order is well defined */
          if (Object.keys(o.relations).every(pool.hasOwnProperty)) {
            return getStepsFromOrder().steps
          } else {
            throw new Error(`${o.model.name} uses undefined relations!`)
          }
        }
        default: {
          const [o, ...os] = remainingOrders

          /* Checks if the order is well defined */
          if (Object.keys(o.relations).every(pool.hasOwnProperty)) {
            const { steps, pool: remainingPool } = getStepsFromOrder()
            return [
              ...steps,
              ...sort(os, [...sortedSteps, ...steps], remainingPool),
            ]
          } else {
            return sort([...os, o], sortedSteps, pool)
          }
        }
      }
      return []
    }

    function getStepsFromOrder(): { steps: Step[]; pool: Pool } {
      // const foo = [
      //   {
      //     order: getStepNumber(pool),
      //     model: o.model,
      //     amount: pool[o.model.name].remainingUnits,
      //     runningNumber: o.amount,
      //     relations: getModelRelations(),
      //   },
      // ]

      return { steps: [], pool: {} }
    }

    /**
     * Calculates the step number from the pool.
     *
     * @param pool
     */
    function getStepNumber(pool: Pool): number {
      return _.sum(Object.values(pool).map(m => m.allUnits - m.remainingUnits))
    }

    /* Triggers the order conversion */
    const steps = sort(orders, [], {})

    return steps
  }

  /**
   * Converts fixtures from steps by creating a pool of available instances
   * and assigning relations to particular types.
   *
   * @param steps
   */
  function getFixturesFromSteps(schema: FakerSchema, steps: Step[]): Fixture[] {
    faker.seed(opts.seed)

    type Pool = Dictionary<ID[]>

    const [fixtures] = _.sortBy(steps, s => s.order).reduce<[Fixture[], Pool]>(
      ([fixtures, pool], step) => {
        const [data, newPool] = getMockDataForStep(pool, step)

        const fixture: Fixture = {
          order: 0,
          id: faker.random.uuid(),
          model: step.model,
          data: data,
        }

        const poolWithFixture = insertInstanceIDIntoPool(
          newPool,
          step.model.name,
          fixture.id,
        )

        return [fixtures.concat(fixture), poolWithFixture]
      },
      [[], {}],
    )

    return fixtures

    /* Helper functions */

    /**
     * Generates mock data from the provided model. Scalars return a mock scalar or
     * list of mock scalars, relations return an ID or lists of IDs.
     */
    function getMockDataForStep(_pool: Pool, step: Step): [FixtureData, Pool] {
      const [finalPool, fixture] = step.model.fields.reduce(
        ([pool, acc], field) => {
          const fieldModel = field.getModel()
          const mock = fallback =>
            withDefault(fallback, schema[step.model.name][fieldModel.name])()

          switch (field.type) {
            case 'ID': {
              const id = mock(faker.random.uuid)

              return [pool, { ...acc, [field.name]: id }]
            }
            case 'String': {
              const string = mock(faker.random.word)

              return [pool, { ...acc, [field.name]: string }]
            }
            case 'Int': {
              const number = mock(faker.random.number)

              return [pool, { ...acc, [field.name]: number }]
            }
            case 'Float': {
              const float = mock(faker.finance.amount)

              return [pool, { ...acc, [field.name]: float }]
            }
            case 'Date': {
              const date = mock(faker.date.past)

              return [pool, { ...acc, [field.name]: date }]
            }
            default: {
              /* Relations */
              if (field.isRelation()) {
                if (field.isList) {
                  const [id, newPool] = getInstanceIDsFromPool(
                    pool,
                    fieldModel.name,
                    step.relations[fieldModel.name].amount,
                  )
                  return [newPool, { ...acc, [field.name]: id }]
                } else {
                  const [id, newPool] = getInstanceIDFromPool(
                    pool,
                    fieldModel.name,
                  )
                  return [newPool, { ...acc, [field.name]: id }]
                }
              }

              /* Custom field mocks */
              if (schema[step.model.name][fieldModel.name]) {
                return schema[step.model.name][fieldModel.name]()
              }

              /* Fallback for unsupported scalars */
              throw new Error(`Unsupported field type "${field.type}".`)
            }
          }
        },
        [_pool, {}],
      )

      return [fixture, finalPool]
    }

    /**
     * Retrieves an ID from the pool and removes its instance.
     */
    function getInstanceIDFromPool(pool: Pool, type: string): [ID, Pool] {
      return [pool[type][0], { ...pool, [type]: pool[type].splice(1) }]
    }

    /**
     * Retrieves an ID from the pool and removes its instance.
     */
    function getInstanceIDsFromPool(
      pool: Pool,
      type: string,
      n: number,
    ): [ID[], Pool] {
      return [pool[type].slice(0, n), { ...pool, [type]: pool[type].splice(n) }]
    }

    /**
     * Inserts an ID into the pool and returns the new pool.
     */
    function insertInstanceIDIntoPool(pool: Pool, type: string, id: ID): Pool {
      return {
        ...pool,
        [type]: [...withDefault([], pool[type]), id],
      }
    }
  }
}
