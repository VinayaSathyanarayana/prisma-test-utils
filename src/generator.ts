import { GeneratorDefinition, GeneratorOptions } from '@prisma/cli'
import { DMMF } from '@prisma/photon/runtime/dmmf-types'
import _ from 'lodash'
import mls from 'multilines'
import * as path from 'path'
import { withDefault } from './utils'
import { VirtualFS, writeToFS } from './vfs'

/**
 * Generates prisma-test-utils library using Prisma Generators.
 * The function presumes there's an existing photon library at an absolute path.
 *
 * @param options
 */
export async function generatePrismaTestUtils(
  options: GeneratorOptions,
): Promise<string> {
  /* Config */
  const outputDir = withDefault('', options.generator.output)
  const photonPath = _.get(
    options.generator.config,
    ['photonPath'],
    '@generated/photon',
  )
  const staticPath = path.resolve(__dirname, './_static/index.js')

  /**
   * The generation process is separated into three parts:
   *  1. We use @zeit/ncc to build the static parts of the library.
   *  2. We wire the functions from the static part with the dynamic properties.
   *  3. We write the files to the file system.
   */

  /* Dynamic files */
  const dmmf = require(photonPath).dmmf as DMMF.Document

  const seedLib = mls`
  | import { seed as staticSeed, SeedOptions } from '${staticPath}'
  |
  | const dmmf = ${JSON.stringify(dmmf)};
  |
  | export default staticSeed(dmmf);
  | export { SeedOptions };
  `
  const poolLib = mls`
  | import { pool as staticPool, PoolOptions } from '${staticPath}'
  |
  | const dmmf = ${JSON.stringify(dmmf)};
  |
  | export default staticPool(dmmf);
  | export { PoolOptions };
  `

  /* Static files */

  const vfs: VirtualFS = {
    'pool.js': poolLib,
    'seed.js': seedLib,
  }

  /**
   * Write files to file system.
   */
  try {
    await writeToFS(outputDir, vfs)
  } catch (err) {
    throw err
  }

  return ''
}

export const generatorDefinition: GeneratorDefinition = {
  prettyName: 'Prisma Test Utils',
  generate: generatePrismaTestUtils,
}
