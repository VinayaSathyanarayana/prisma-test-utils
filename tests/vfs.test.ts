import * as fs from 'fs'
import ml from 'multilines'
import * as os from 'os'
import * as path from 'path'
import { promisify } from 'util'
import { VirtualFS, writeToFS, copyVFS, compileVFS } from '../src/vfs'
import { ModuleKind, ScriptTarget } from 'typescript'

const mkdir = promisify(fs.mkdir)
const writeFile = promisify(fs.writeFile)

describe('virtual file system:', () => {
  test('compiles TypeScript files correctly', async () => {
    const root = path.join(os.tmpdir(), `./vfs-compile-test-${Math.random()}`)
    const vfs: VirtualFS = {
      [path.join(root, 'foo.ts')]: ml`
      | import { sum } from './bar.ts'
      | console.log(sum(1,2))
      `,
      [path.join(root, 'bar.ts')]: ml`
      | export function sum(a: number, b: number): number {
      |   return a + b
      | }
      `,
      'qux.js': ml`
      | console.log("Hey")
      `,
    }

    const compiledVFS = await compileVFS(vfs, {
      module: ModuleKind.CommonJS,
      target: ScriptTarget.ES2016,
      lib: ['lib.esnext.d.ts', 'lib.dom.d.ts'],
      declaration: true,
      sourceMap: true,
      suppressOutputPathCheck: false,
    })

    expect(Object.values(compiledVFS)).toMatchSnapshot()
  })

  test('copies virtual file system correctly', async () => {
    const root = path.join(os.tmpdir(), `./copy-vfs-tests-${Math.random()}`)
    const dest = path.join(os.tmpdir(), `./copy-vfs-tests-d-${Math.random()}`)

    /* Random test files. */
    const files = [
      '%root/files/file_1',
      '%root/files/file_2',
      '%root/file_3',
    ].map(filePath => ({
      path: filePath.replace('%root', root),
      copyPath: filePath.replace('%root', dest),
      content: `file: ${Math.random()}`,
    }))

    /* Write files to the FS */

    await Promise.all(
      files.map(async file => {
        await mkdir(path.dirname(file.path), { recursive: true })
        await writeFile(file.path, file.content)
      }),
    )

    /* ViftualFS representation. */

    const vfs: VirtualFS = {
      [path.join(dest, './files')]: path.join(root, './files'),
      [path.join(dest, './file_3')]: path.join(root, './file_3'),
    }

    await copyVFS(vfs)

    /* Tests. */

    await Promise.all(
      files.map(async file => {
        expect(fs.readFileSync(file.copyPath).toString()).toEqual(file.content)
      }),
    )
  })

  test('writes to fs', async () => {
    /* Random test files. */
    const files = [0, 1, 2, 3, 4].map(file => ({
      path: path.join(os.tmpdir(), `./write-vfs-tests-${Math.random()}`),
      content: `testFile: ${file} random number: ${Math.random()}`,
    }))

    /* VirtualFS representation. */
    const vfs: VirtualFS = files.reduce(
      (acc, file) => ({
        ...acc,
        [file.path]: file.content,
      }),
      {},
    )

    await writeToFS(vfs)

    /* Tests. */

    await Promise.all(
      files.map(file => {
        expect(fs.readFileSync(file.path).toString()).toEqual(file.content)
      }),
    )
  })
})
