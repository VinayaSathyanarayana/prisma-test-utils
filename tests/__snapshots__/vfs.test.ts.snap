// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`virtual file system: compiles TypeScript files correctly 1`] = `
Object {
  "/var/folders/46/ghtsn_bd0m3f7g2kcscx_jqc0000gn/T/vfs-compile-test-0.14287605890623367/bar.d.ts": "export declare function sum(a: number, b: number): number;
",
  "/var/folders/46/ghtsn_bd0m3f7g2kcscx_jqc0000gn/T/vfs-compile-test-0.14287605890623367/bar.js": "\\"use strict\\";
Object.defineProperty(exports, \\"__esModule\\", { value: true });
function sum(a, b) {
    return a + b;
}
exports.sum = sum;
//# sourceMappingURL=bar.js.map",
  "/var/folders/46/ghtsn_bd0m3f7g2kcscx_jqc0000gn/T/vfs-compile-test-0.14287605890623367/bar.js.map": "{\\"version\\":3,\\"file\\":\\"bar.js\\",\\"sourceRoot\\":\\"\\",\\"sources\\":[\\"bar.ts\\"],\\"names\\":[],\\"mappings\\":\\";;AAAA,SAAgB,GAAG,CAAC,CAAS,EAAE,CAAS;IACtC,OAAO,CAAC,GAAG,CAAC,CAAA;AACd,CAAC;AAFD,kBAEC\\"}",
  "/var/folders/46/ghtsn_bd0m3f7g2kcscx_jqc0000gn/T/vfs-compile-test-0.14287605890623367/foo.d.ts": "export {};
",
  "/var/folders/46/ghtsn_bd0m3f7g2kcscx_jqc0000gn/T/vfs-compile-test-0.14287605890623367/foo.js": "\\"use strict\\";
Object.defineProperty(exports, \\"__esModule\\", { value: true });
const bar_ts_1 = require(\\"./bar.ts\\");
console.log(bar_ts_1.sum(1, 2));
//# sourceMappingURL=foo.js.map",
  "/var/folders/46/ghtsn_bd0m3f7g2kcscx_jqc0000gn/T/vfs-compile-test-0.14287605890623367/foo.js.map": "{\\"version\\":3,\\"file\\":\\"foo.js\\",\\"sourceRoot\\":\\"\\",\\"sources\\":[\\"foo.ts\\"],\\"names\\":[],\\"mappings\\":\\";;AAAA,qCAA8B;AAC9B,OAAO,CAAC,GAAG,CAAC,YAAG,CAAC,CAAC,EAAC,CAAC,CAAC,CAAC,CAAA\\"}",
  "qux.js": "console.log(\\"Hey\\")",
}
`;
